cleandart/react-dart
# New JS interop, performance/lifecycle improvements

# Description
Dart 1.13 introduced the new [`js` package](https://pub.dartlang.org/packages/js), which provides an alternate Dart-JS interop that is faster than `dart:js` interop.

react-dart currently uses `dart:js` interop, and could benefit from performance improvements in `package:js`.

This PR upgrades react-dart to use `package:js`, and also includes other changes that improve performance and increase parity with React JS.

## Performance data:
Performance tests were run on an earlier version of these changes, as well as for the current react-dart and React JS.

Full results available [in this doc](https://docs.google.com/spreadsheets/d/1GB8QZZGlgFNw4gx6AuRA242Ardfei0sDAYRcHv_osLU/edit?usp=sharing
). Includes data for:
* "unit"-style rendering performance: (https://github.com/greglittlefield-wf/react-dart_perf_unit)
* functional-style frame rate performance: (https://github.com/greglittlefield-wf/react-dart_perf/tree/new_react-dart_interop)

_See https://github.com/greglittlefield-wf/react-dart_perf_unit/blob/master/README.md#improvement-details for more details on which performance improvements were included while testing._

Note: these results also include numbers for a few potential dart2js optimizations (more details in the link above). I will follow up this PR with links to dart-lang/sdk GitHub issues regarding these potential optimizations.

#### Summary of results
###### react-dart Initial Rendering Time (in terms of pure JS time):
* current: 14.3x
* optimized: 5.7x
* optimized + dart2js tweaks: 3.5x

###### react-dart Frame Rate (% of pure JS React rate):
* current: 21.5%
* optimized: 46.5%
* optimized + dart2js tweaks: 60.9%

# Changes
##### Performance improvements:
* Use new `package:js` Dart-JS interop in `react_client` and `react_test_utils`.
    * Update SDK constraint in `pubspec.yaml`.
* Optimize event handler conversion by using a Map instead of if-else.
* Lifecycle-related changes: the following React component lifecycle changes were made for the sake of performance as well as better parity with React JS:
    * Remove `rootNode` argument from `componentDidMount`/`componentDidUpdate`.
        * Prevents unnecessary calls to `findDOMNode` and creation of dart2js interceptors for Elements, which affects performance.
    * Only call `getDefaultProps()` once and cache the result.
    * Merge default props only once, in the component factory.
    * Cache `nextProps` on the component to prevent excessive calls to `_getNextProps`.

##### Other changes
* Use a Dart class (`ReactDartComponentInternal`) instead of a Map for managing react-dart component internals.
* Store internal object in the JS props at the `internal` key, instead of `__internal__`.
    * `package:js` doesn't allow you to proxy fields starting with underscores without making them private.
* Don't expose `key`/`ref` in a rendered component's props (React JS started doing this [in 0.12](https://facebook.github.io/react/blog/2014/10/16/react-v0.12-rc1.html#breaking-change-key-and-ref-removed-from-this.props)).
* Add `test/lifecycle_test` to verify the lifecycle behavior of Dart components.
* `shouldComponentUpdate`, `componentDidUpdate`, and `componentWillUnmount` are no longer skippable via `skipMethods`.

### Breaking Changes
From the changes mentioned above, the following are breaking:
* react-dart now requires a Dart SDK of `>=1.13.0`.
* `componentDidMount`/`componentDidUpdate` no longer have the `rootNode` argument. 
* `ReactElement`s now always have default props merged in.
* `package:js` interop-related:
    * React element instances are now of type `ReactElement`, not `JsObject`.
        * (These are the objects returned when invoking a component factory.) 
    * React component instances are now of type `ReactComponent`, not `JsObject`.
        * (These are the objects returned by `ref` and `react.render`.)
    * React component classes are now of type `ReactClass`, not `JsFunction`
    * JS props are now `package:js` interop objects, not `JsObject`.
* `props['children']` is now a non-growable `List`.
* Internals:
    * A Dart class (`ReactDartComponentInternal`) is used instead of a Map for managing react-dart component internals.
    * The react-dart "internal" object is stored in the JS props at the `internal` key, instead of `__internal__`.

# Testing
* In both Dartium and dart2js:
    * Verify that all tests in `test/` pass.
    * Verify that all examples in `example/test/` work as expected.

---

@hleumas @trentgrover-wf @evanweible-wf @maxwellpeterson-wf @timmccall-wf @todbachman-wf @aaronlademann-wf @clairesarsam-wf @jacehensley-wf @joelleibow-wf
